components/core/tests/test-ffi_KeyValuePairLogEvent.cpp

```diff
@@ -0,0 +1,529 @@
+        REQUIRE((value.get_immutable_view<Type>() == typed_value));
+        REQUIRE((std::is_same_v<Type, decltype(value.get_immutable_view<Type>())>));
+    } else {
+        REQUIRE_THROWS(value.get_immutable_view<value_bool_t>());
+    }
+
+    if constexpr (std::is_same_v<string, Type>) {
+        REQUIRE((value.get_immutable_view<Type>() == typed_value));
+        REQUIRE((std::is_same_v<std::string_view, decltype(value.get_immutable_view<Type>())>));
+    } else {
+        REQUIRE_THROWS(value.get_immutable_view<string>());
+    }
+
+    if constexpr (std::is_same_v<EightByteEncodedTextAst, Type>) {
+        REQUIRE((value.get_immutable_view<Type>() == typed_value));
+        REQUIRE((std::is_same_v<
+                 EightByteEncodedTextAst const&,
+                 decltype(value.get_immutable_view<Type>())>));
+    } else {
+        REQUIRE_THROWS(value.get_immutable_view<EightByteEncodedTextAst>());
+    }
+
+    if constexpr (std::is_same_v<FourByteEncodedTextAst, Type>) {
+        REQUIRE((value.get_immutable_view<Type>() == typed_value));
+        REQUIRE((std::is_same_v<
+                 FourByteEncodedTextAst const&,
+                 decltype(value.get_immutable_view<Type>())>));
+    } else {
+        REQUIRE_THROWS(value.get_immutable_view<FourByteEncodedTextAst>());
+    }
+}
+
+auto insert_invalid_node_id_value_pairs_with_node_type_errors(
+        SchemaTree const& schema_tree,
+        SchemaTreeNode::id_t node_id,
+        KeyValuePairLogEvent::NodeIdValuePairs& invalid_node_id_value_pairs
+) -> void {
+    REQUIRE((node_id < schema_tree.get_size()));
+    auto const node_type{schema_tree.get_node(node_id).get_type()};
+    if (SchemaTreeNode::Type::Int != node_type) {
+        invalid_node_id_value_pairs.emplace(node_id, Value{static_cast<value_int_t>(0)});
+    }
+    if (SchemaTreeNode::Type::Float != node_type) {
+        invalid_node_id_value_pairs.emplace(node_id, Value{static_cast<value_float_t>(0.0)});
+    }
+    if (SchemaTreeNode::Type::Bool != node_type) {
+        invalid_node_id_value_pairs.emplace(node_id, Value{static_cast<value_bool_t>(false)});
+    }
+    if (SchemaTreeNode::Type::Str != node_type) {
+        invalid_node_id_value_pairs.emplace(node_id, Value{static_cast<string>("Test")});
+        if (SchemaTreeNode::Type::UnstructuredArray != node_type) {
+            invalid_node_id_value_pairs.emplace(
+                    node_id,
+                    Value{get_encoded_text_ast<four_byte_encoded_variable_t>(cStringToEncode)}
+            );
+            invalid_node_id_value_pairs.emplace(
+                    node_id,
+                    Value{get_encoded_text_ast<eight_byte_encoded_variable_t>(cStringToEncode)}
+            );
+        }
+    }
+    if (SchemaTreeNode::Type::Obj != node_type) {
+        invalid_node_id_value_pairs.emplace(node_id, std::nullopt);
+        invalid_node_id_value_pairs.emplace(node_id, Value{});
+    }
+}
+
+auto clone_value(Value const& value) -> Value {
```

Comment:
Hm, I missed this in the previous PRs but it seems like we should just have copy constructors for `Value` and `EncodedTextAst`. It feels like an overoptimization to force people to never use copy construction unless they write a method manually like this.

Reply:
What's bad about copying a `Value`? Is it not similar to copying a `string`?

Reply:
Copying string and `EncodedTextAst` can be expensive in deserialization. With copy disabled, we are more confident that there's no copy overhead in deserialization.

Reply:
Adding move constructor for `EncodedTextAst` and `Value`